File: c_sharp/src/org/ldk/structs/Option_AddressZ.cs
using org.ldk.impl;
using org.ldk.enums;
using org.ldk.util;
using System;

namespace org { namespace ldk { namespace structs {

/**
 * An enum which can either contain a crate::c_types::Address or not
 */
public class Option_AddressZ : CommonBase {
	protected Option_AddressZ(object _dummy, long ptr) : base(ptr) { }
	~Option_AddressZ() {
		if (ptr != 0) { bindings.COption_AddressZ_free(ptr); }
	}

	internal static Option_AddressZ constr_from_ptr(long ptr) {
		long raw_ty = bindings.LDKCOption_AddressZ_ty_from_ptr(ptr);
		switch (raw_ty) {
			case 0: return new Option_AddressZ_Some(ptr);
			case 1: return new Option_AddressZ_None(ptr);
			default:
				throw new ArgumentException("Impossible enum variant");
		}
	}

	/** A Option_AddressZ of type Some */
	public class Option_AddressZ_Some : Option_AddressZ {
		public org.ldk.structs.Address some;
		internal Option_AddressZ_Some(long ptr) : base(null, ptr) {
			long some = bindings.LDKCOption_AddressZ_Some_get_some(ptr);
			Address some_conv = new Address(null, some);
			this.some = some_conv;
		}
	}
	/** A Option_AddressZ of type None */
	public class Option_AddressZ_None : Option_AddressZ {
		internal Option_AddressZ_None(long ptr) : base(null, ptr) {
		}
	}
	/**
	 * Constructs a new COption_AddressZ containing a crate::c_types::Address
	 */
	public static org.ldk.structs.Option_AddressZ some(org.ldk.structs.Address o) {
		long ret = bindings.COption_AddressZ_some(o.ptr);
		GC.KeepAlive(o);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_AddressZ ret_hu_conv = org.ldk.structs.Option_AddressZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Constructs a new COption_AddressZ containing nothing
	 */
	public static org.ldk.structs.Option_AddressZ none() {
		long ret = bindings.COption_AddressZ_none();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_AddressZ ret_hu_conv = org.ldk.structs.Option_AddressZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

}
} } }

File: src/main/java/org/ldk/structs/Option_AddressZ.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;
import java.lang.ref.Reference;
import javax.annotation.Nullable;


/**
 * An enum which can either contain a crate::c_types::Address or not
 */
@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class Option_AddressZ extends CommonBase {
	private Option_AddressZ(Object _dummy, long ptr) { super(ptr); }
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		super.finalize();
		if (ptr != 0) { bindings.COption_AddressZ_free(ptr); }
	}
	static Option_AddressZ constr_from_ptr(long ptr) {
		bindings.LDKCOption_AddressZ raw_val = bindings.LDKCOption_AddressZ_ref_from_ptr(ptr);
		if (raw_val.getClass() == bindings.LDKCOption_AddressZ.Some.class) {
			return new Some(ptr, (bindings.LDKCOption_AddressZ.Some)raw_val);
		}
		if (raw_val.getClass() == bindings.LDKCOption_AddressZ.None.class) {
			return new None(ptr, (bindings.LDKCOption_AddressZ.None)raw_val);
		}
		assert false; return null; // Unreachable without extending the (internal) bindings interface
	}

	/**
	 * When we're in this state, this COption_AddressZ contains a crate::c_types::Address
	 */
	public final static class Some extends Option_AddressZ {
		public final org.ldk.structs.Address some;
		private Some(long ptr, bindings.LDKCOption_AddressZ.Some obj) {
			super(null, ptr);
			long some = obj.some;
			Address some_conv = new Address(null, some);
			this.some = some_conv;
		}
	}
	/**
	 * When we're in this state, this COption_AddressZ contains nothing
	 */
	public final static class None extends Option_AddressZ {
		private None(long ptr, bindings.LDKCOption_AddressZ.None obj) {
			super(null, ptr);
		}
	}
	/**
	 * Constructs a new COption_AddressZ containing a crate::c_types::Address
	 */
	public static Option_AddressZ some(org.ldk.structs.Address o) {
		long ret = bindings.COption_AddressZ_some(o.ptr);
		Reference.reachabilityFence(o);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_AddressZ ret_hu_conv = org.ldk.structs.Option_AddressZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Constructs a new COption_AddressZ containing nothing
	 */
	public static Option_AddressZ none() {
		long ret = bindings.COption_AddressZ_none();
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.Option_AddressZ ret_hu_conv = org.ldk.structs.Option_AddressZ.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

}
